perllib/FixMyStreet/App/Controller/Admin.pm

```diff
@@ -1149,14 +1155,33 @@ sub user_edit : Path('user_edit') : Args(1) {
         $user->is_superuser( ( $c->user->is_superuser && $c->get_param('is_superuser') ) || 0 );
         # Superusers can set from_body to any value, but other staff can only
         # set from_body to the same value as their own from_body.
-        if ($c->user->is_superuser) {
+        if ( $c->user->is_superuser ) {
             $user->from_body( $c->get_param('body') || undef );
-        } elsif ($c->get_param('body') eq $c->user->from_body->id) {
+        } elsif ( $c->user->has_permission_to('user_assign_body', $c->user->from_body->id ) &&
+                  $c->get_param('body') && $c->get_param('body') eq $c->user->from_body->id ) {
             $user->from_body( $c->user->from_body );
         } else {
             $user->from_body( undef );
         }
 
+        if (!$user->from_body) {
+            # Non-staff users aren't allowed any permissions
+            $user->user_body_permissions->delete_all;
+        } elsif ($c->stash->{available_permissions} && $c->user->has_permission_to('user_manage_permissions', $user->from_body->id)) {
+            my @all_permissions = map { map { $_ } keys $_ } values $c->stash->{available_permissions};
```

Comment:
`available_permissions` is a nested hashref and I wanted to flatten it out to a simple list of permission names, the inner map seemed like the simplest way to do that. Is there a better approach?